source/common/http/conn_pool_base.cc

```diff
@@ -1,16 +1,66 @@
 #include "common/http/conn_pool_base.h"
 
+#include "common/network/transport_socket_options_impl.h"
+#include "common/runtime/runtime_features.h"
 #include "common/stats/timespan_impl.h"
 #include "common/upstream/upstream_impl.h"
 
 namespace Envoy {
 namespace Http {
+Network::TransportSocketOptionsSharedPtr
+wrapTransportSocketOptions(Network::TransportSocketOptionsSharedPtr transport_socket_options,
+                           Protocol protocol) {
+  if (!Runtime::runtimeFeatureEnabled("envoy.reloadable_features.http_default_alpn")) {
+    return transport_socket_options;
+  }
+
+  // TODO(reviewers): Given that we pass along the ALPN fallback over the same public interface that
+  // is provided as part of the LBContext, its possible that the ALPN fallback has already been set
```

Comment:
Do you mean the override has already been set? See my other comment. I'm confused. :(

Reply:
The way the `LoadBalancerContext` interface works today means that callers (e.g. the router filter) are able to pass along a `TransportSocketOptionsSharedPtr` that is used to configure the upstream connection. This gets passed through to the conn manager, which uses it to determine which conn pool to use, and then further on to `createConnection` on the host to actually create the connection.

The problem I'm highlighting here is that because the same object is passed through this whole chain, modifying these options in the middle of it (during conn pool creation) makes for an awkward API since it means that value for `alpn_fallback` might have been set at the start of the chain (e.g. by the router). The code right now just overwrites any such value when the transport socket options are passed through the conn pool.

One of the options here would be to have multiple transport socket option types we use, so that we're not exposing all of these options through the LoadBalancerContext interface. At the same time, putting this configuration into the `TransportSocketOptions` object seemed like the natural place to put it. 

The API complexity with fallbacks and overrides comes from the fact that the configured ALPN is buried deep into the transport socket, which makes it a bit tricky to have a straight up "which ALPN should I use" API. Instead, in this PR there's an override which always takes precedence (existing API) and a new fallback option that is selected when no ALPN was configured through the transport socket config. 

This all feels more complex than it ought to be so I'm open for suggestions

Reply:
How is the override functionality used today? Can we just piggy back on that? I think the override vs. fallback is what is confusing me right now.

Reply:
It's set when `Network::ApplicationProtocols` is set on the filter state: https://github.com/envoyproxy/envoy/blob/master/source/common/network/transport_socket_options_impl.cc#L48-L53

This isn't used in tree anywhere, but it seems like Istio uses this to accomplish exactly what we're trying to do here, to set the ALPN to match the upstream protocol: https://github.com/istio/proxy/blob/master/src/envoy/http/alpn/alpn_filter.cc#L76-L85

I guess this means that one option would be to bring such a filter into the main repo and tell people that want this behavior to use that.

The problem with using the ALPN override functionality to do this is that we'd stomp on the ALPN settings set on the TLS context. Maybe we should be exposing the configured ALPN from the `TransportSocketFactory` (which can be accessed from the `HostDescription`), which would let us query for the configured ALPN before setting the override?

Reply:
> I guess this means that one option would be to bring such a filter into the main repo and tell people that want this behavior to use that.

In an optimal world I think we should have this behavior be built-in, since it's such a common point of misconfiguration.

> The problem with using the ALPN override functionality to do this is that we'd stomp on the ALPN settings set on the TLS context. Maybe we should be exposing the configured ALPN from the TransportSocketFactory (which can be accessed from the HostDescription), which would let us query for the configured ALPN before setting the override?

OK I see. Per above I think it's OK to have these different cases, but I found it confusing from reading. Can you add more comments. Also, from my initial read through it wasn't clear to me how the override and fallback stuff interacts with each other (additive, overwrites, etc.). Can you add more comments there also and I can look again?